src/printer.js

```diff
@@ -2032,27 +2032,68 @@ function printJSXElement(path, options, print) {
     var children = [];
 
     path.map(
-      function(childPath) {
+      function(childPath, i) {
         var child = childPath.getValue();
+        let lastIdx = children.length - 1;
+        let lastChild = children[lastIdx];
 
         if (
           namedTypes.Literal.check(child) && typeof child.value === "string"
         ) {
           if (/\S/.test(child.value)) {
             const beginBreak = child.value.match(/^\s*\n/);
             const endBreak = child.value.match(/\n\s*$/);
+            const beginSpace = child.value.match(/^\s+/);
+            const endSpace = child.value.match(/\s+$/);
+
+            if (beginBreak) {
+              children.push(hardline);
+            } else if (beginSpace) {
+              let prev = util.getLast(children);
+              if (prev && prev.type === "line") {
+                // TODO ifBreak with else
+                children.push("{' '}", softline);
+              } else {
+                children.push(" ");
+              }
+            }
+
+            children.push(child.value.replace(/^\s+|\s+$/g, ""));
+
+            if (endBreak) {
+              children.push(hardline);
+            } else if (endSpace) {
+              children.push(" ");
+            }
 
-            children.push(
-              beginBreak ? hardline : "",
-              child.value.replace(/^\s+|\s+$/g, endBreak ? "" : " "),
-              endBreak ? hardline : ""
-            );
           } else if (/\n/.test(child.value)) {
             children.push(hardline);
+          } else if (/\s/.test(child.value)) {
+            children.push(" ");
+            children.push(softline);
           }
         } else {
           children.push(print(childPath));
         }
+
+        if (
+          (child.openingElement && child.openingElement.selfClosing) ||
+          (child.closingElement && child.closingElement.type === 'JSXClosingElement')
+        ) {
+          let nextChild = n.children[i + 1];
+
+          if (!nextChild || nextChild.type === 'JSXElement') {
+            children.push(softline);
+          } else if (!nextChild.value.match(/\S/) && !nextChild.value.match(/\n/)) {
+            // children.push(softline);
+          }
+        }
+
+        if (typeof lastChild === "string" && lastChild.match(/\s$/)) {
+          if (children[lastIdx + 1].type === "line") {
+            children[lastIdx] = lastChild.replace(/\s$/, "{' '}");
```

Comment:
Sorry, missed this comment. I thought JSX removes whitespace at the beginning and end of a line? What is this whitespace significant? Can you paste some JSX code to help me get the context of this?

I'd prefer to avoid the printer being aware of JSX, but once I understand why this is needed we can figure something out.